### PR TITLE
Use easy.to_list and easy.to_tuple

### PR DESCRIPTION
**Pull request by Gouvernathor**
replaces #3944 
I chose to test for str and not basestring, so that the behavior is exactly the same on py2 and py3, but that's arguable (now it only applies to unicode, to be clear).
The default value of `copy` is also arguable.
Another or_none parameter could be added, to avoid returning [None] in some cases.

### CODE REVIEW

File: renpy/easy.py [hunk 1]
```diff
@@ -258,3 +263,22 @@ def split_properties(properties, *prefixes):
...
+def to_list(value, copy=False):
+    """
+    If the value is an iterable, turns it into a list, otherwise wraps it into one.
+    If a list is provided and `copy` is True, a new list will be returned.
+    """
+    if isinstance(value, Iterable) and not isinstance(value, str):
+        if (not copy) and isinstance(value, list):
+            return value
+        return list(value)
+    return [value]
```
Comment (mal) [inline renpy/easy.py:276]:
```suggestion
    if isinstance(value, list):
        return list(value) if copy else value
        
    if not isinstance(value, str) and isinstance(value, Iterable):
        return list(value)

    return [value]
```

File: renpy/easy.py [hunk 2]
```diff
@@ -258,3 +263,22 @@ def split_properties(properties, *prefixes):
...
+
+def to_tuple(value):
+    """
+    Same as to_list, but with tuples.
+    """
+    if isinstance(value, Iterable) and not isinstance(value, str):
+        return tuple(value)
+    return (value,)
```
Comment (mal) [inline renpy/easy.py:284]:
```suggestion
    if isinstance(value, tuple):
        return value

    if not isinstance(value, str) and isinstance(value, Iterable):
        return tuple(value)

    return (value,)
```

File: renpy/display/behavior.py [hunk 3]
```diff
@@ -222,8 +222,7 @@ def map_event(ev, keysym):
 
         return False
 
-    if isinstance(keysym, list):
-        keysym = tuple(keysym)
+    keysym = renpy.easy.to_tuple(keysym)
```
Comment (mal) [inline renpy/display/behavior.py:225]: There doesn't appear to be any expectation here that the `keysym` argument wouldn't already be an `Iterable` of some kind, and I don't think we should be accidentally adding more expensive code here to allow it. These are internal functions that already work, in contrast to the proposed changes to `layeredimage` and `at_list` code which is dealing directly with creator input and providing a direct benefit.

File: renpy/display/behavior.py [hunk 4]
```diff
@@ -240,8 +239,7 @@ def map_keyup(ev, keysym):
         if (keysym in ev.eventnames) and ev.up:
             return True
 
-    if isinstance(keysym, list):
-        keysym = tuple(keysym)
+    keysym = renpy.easy.to_tuple(keysym)
```
Comment (mal) [inline renpy/display/behavior.py:242]: Same opinion here as with the proposed `map_event` change.

**Review by mal — COMMENTED** [review]
I think it would be good in the proposed helpers to optimize for the expected cases of the type being asked for, then `str`, and then `Iterable`; short-circuiting as necessary to minimise the number of checks. Somewhat more subjectively, I'd consider that approach to be much easier to reason about when reading the code. With that in mind, for your consideration:

File: renpy/display/behavior.py (hunk 3 above)
Comment (Gouvernathor) [reply renpy/display/behavior.py:225]: I was under the impression that the keysym directly came from the user, even though the function itself is not public ? I don't know, but if non-list iterables of keysyms are currently accepted as well as lists, then we can surely kick it off.

File: renpy/easy.py (hunk 1 above)
Comment (Gouvernathor) [reply renpy/easy.py:276]: Yes, real instance checks are faster than abc checks, let's do them first.

File: renpy/display/behavior.py (hunk 3 above)
Comment (mal) [reply renpy/display/behavior.py:225]: I didn't get the rationale quite right, but did, I believe, get the right conclusion. The correct explanation is that it comes down to how `compile_event` works, and that converting non-iterables into iterables here needlessly creates more work in that function, and results in more complex dynamic code being generated there.

**Pull request merged**